src/SSHDebugPS/RawShell.cs

```diff
@@ -0,0 +1,295 @@
+                    task.Wait(token);
+                    //if (task.Result == 0)
+                    //{
+                    //    Closed?.Invoke(this, 0);
+                    //    Close();
+                    //    return;
+                    //}
+                    if (task.Result > 0)
+                        action(new string(buffer, 0, task.Result));
+                }
+            }
+            catch (Exception)
+            {
+                ErrorOccured?.Invoke(this, null);
+                Close();
+                // close correctly?
+            }
+        }
+
+        private void Close()
+        {
+            Dispose();
+        }
+
+        public void Dispose()
+        {
+            if (!_isClosed)
+            {
+                _isClosed = true;
+                _cancellationSource.Cancel();
+                if (_localProcess != null)
+                    _localProcess.Exited -= OnProcessExited;
+                _localProcess?.Close();
+
+                _stdoutWriter?.Close();
+                _processReader?.Close();
+                _processError?.Close();
+
+                _localProcess = null;
+                _stdoutWriter = null;
+                _processReader = null;
+                _processError = null;
+            }
+        }
+
+        public void Write(string text)
+        {
+            _stdoutWriter.Write(text);
+            _stdoutWriter.Flush();
+        }
+
+        public void WriteCommandStart(string startCommand)
+        {
+            WriteLine(startCommand);
+        }
+
+        public void WriteLine(string text)
+        {
+            _stdoutWriter.WriteLine(text);
+            _stdoutWriter.Flush();
+        }
+
+        private void OnProcessExited(object sender, EventArgs e)
+        {
+            Closed?.Invoke(this, _localProcess.ExitCode);
+        }
+    }
+
+    /// <summary>
+    ///  Shell that uses a remote Connection to send commands and receive input/output
+    /// </summary>
+    internal class RawRemoteShell : IRawShell, IDebugUnixShellCommandCallback
+    {
+        private IDebugUnixShellAsyncCommand _asyncCommand;
+        private bool _isRunning;
+
+        public RawRemoteShell(string command, string arguments, Connection remoteConnection)
+        {
+            string commandText = string.Concat(command, " ", arguments);
+            remoteConnection.BeginExecuteAsyncCommand(commandText, runInShell: false, callback: this, asyncCommand: out _asyncCommand);
+            _isRunning = true;
+        }
+
+        public event EventHandler<string> OutputReceived;
+        public event EventHandler<int> Closed;
+        public event EventHandler ErrorOccured;
+
+        public void Dispose()
+        {
+            if (_isRunning)
+            {
+                _asyncCommand.Abort();
+                _isRunning = false;
+            }
+        }
+
+        public void Write(string text)
+        {
+            if (_isRunning)
```

Comment:
>if (_isRunning) [](start = 12, length = 15)

Could be factored out into an `EnsureRunning` method so you don't have duplicated logic in each method body